File: garden/seeds/java/app/src/main/java/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/util/RxUtils.java
package com.{{company_name}}.android.{{app_package_name_prefix}}.util;

import com.{{company_name}}.android.{{app_package_name_prefix}}.activity.BaseActivity;
import com.{{company_name}}.android.{{app_package_name_prefix}}.fragment.BaseFragment;

import rx.Observable;
import com.trello.rxlifecycle.RxLifecycle;

/**
 * Utility methods for working with RxJava observables
 */
public class RxUtils {

    RxUtils() {
        // No instances..
    }

    /**
     * Bind the given observable to the activity.
     * <p/>
     * This method ensures that observables are unsubscribed from at the correct time + dont
     * call back to the activity in an invalid state
     *
     * @param activity   The activity to bind to
     * @param observable The observable to bind
     */
    public static <T> Observable<T> bindActivity(BaseActivity activity, Observable<T> observable) {
        return observable.compose(RxLifecycle.<T>bindActivity(activity.lifecycle()));
    }

    /**
     * Bind the given observable to the fragment.
     * <p/>
     * This method ensures that observables are unsubscribed from at the correct time + dont
     * call back to the fragment in an invalid state
     *
     * @param fragment   The fragment to bind to
     * @param observable The observable to bind
     */
    public static <T> Observable<T> bindFragment(BaseFragment fragment, Observable<T> observable) {
        return observable.compose(RxLifecycle.<T>bindFragment(fragment.lifecycle()));
    }

}

File: garden/seeds/kotlin/app/src/main/kotlin/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/util/RxUtils.kt
package com.{{company_name}}.android.{{app_package_name_prefix}}.util

import com.{{company_name}}.android.{{app_package_name_prefix}}.activity.BaseActivity
import com.{{company_name}}.android.{{app_package_name_prefix}}.fragment.BaseFragment
import com.{{company_name}}.android.{{app_package_name_prefix}}.mvp.view.MvpView
import com.trello.rxlifecycle.RxLifecycle
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Func1
import rx.schedulers.Schedulers

public fun<T> Observable<T>.bind(view: MvpView) : Observable<T> {
    if (view is BaseActivity) {
        return bind(view)
    } else if (view is BaseFragment) {
        return bind(view)
    } else {
        return this
    }
}

private fun <T> Observable<T>.bind(activity : BaseActivity) : Observable<T> {
    return this.compose<T>(RxLifecycle.bindActivity(activity.lifecycle()));
}

private fun <T> Observable<T>.bind(fragment : BaseFragment) : Observable<T> {
    return this.compose<T>(RxLifecycle.bindFragment(fragment.lifecycle()));
}

fun <T> Observable<T>.observeOnMainThread() : Observable<T> {
    return this.observeOn(AndroidSchedulers.mainThread())
}

fun <T> Observable<T>.subscribeOnIoThread() : Observable<T> {
    return this.observeOn(Schedulers.io())
}

fun <T> Observable<T>.filterNulls() : Observable<T> {
    return this.filter { it != null }
}

fun <T, S : List<T>> Observable<S>.flatMapList() : Observable<T> {
    return flatMap { Observable.from(it) }
}